### ISSUE TITLE
[Windows] Error message from pgsql server is not decoded correctly when the PGconn status is ConnStatus.BAD

### ISSUE THREAD

**Issue by lyq1996** [psycopg/psycopg#194]
First of all, sorry for my poor english..

I am writing a library for pgsql to crack the password of a database user from a password hash, and I use psycopg3 to verify that the password is correct. I found that when the connection is not established and the encoding is not utf-8, psycopg3 does not seem to correctly decode the error message from the server.

My pgsql server runs on windows 10, and the server version is 13.3 and encoding GBK(CP936) for messages displayed.
postgresql.conf:
```
# These settings are initialized by initdb, but they can be changed.
lc_messages = 'Chinese (Simplified)_China.936'			# locale for system error message
					# strings
lc_monetary = 'Chinese (Simplified)_China.936'			# locale for monetary formatting
lc_numeric = 'Chinese (Simplified)_China.936'			# locale for number formatting
lc_time = 'Chinese (Simplified)_China.936'				# locale for time formatting
```

For example, If the user's password was wrong or datebase for login doesn't existed, the obj.status will be ConnStatus.BAD(https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/pq/misc.py#L64), and then the msg will decoded as utf-8 by default and cause an error, then the part of msg which can not be decoded will be replaced by '\ufffd'. 

Here is the exmaple code:
```
# coding=utf-8
import time
import psycopg

try:
    conn = psycopg.connect("user=user1 password='123456' host='127.0.0.1' port='5432' client_encoding='GBK'")
except psycopg.Error as ex:
    print(str(ex))
```
The error message should be "connection failed: 致命错误:  数据库 "user1" 不存在" which meaning "FATAL:  database "user1" does not exist\n".

![screenshot](https://user-images.githubusercontent.com/20365480/147800928-7fc3c3cb-d0cc-424e-8a89-e1d8f14bcb1e.png)

When the connection established, the error message can be decoded correctly according to the given client_encoding:
```
# coding=utf-8
import time
import psycopg

try:
    conn = psycopg.connect("user=postgres password='123456' host='127.0.0.1' port='5432' client_encoding='GBK'")
    conn.execute("""
                select * from qwertsdnmsdjffks;)
                """)
    conn.commit()
except psycopg.Error as ex:
    print(str(ex))
```
![screenshot2](https://user-images.githubusercontent.com/20365480/147801478-a8ee8829-0ff8-4b27-879b-0d4603deb7c5.png)

Hope you can confirm and fix, thanks a lot.

**Comment by dvarrazzo** [comment]
I assume  that, when that error happens on connection, we don't know the server encoding yet. 

I assume that you assume that psycopg would use the encoding specified in the connection string. It doesn't: that encoding is passed to the server and then psycopg reads it back from the connection, but that doesn't happen when the connection is returned bad. I can see if it is not too difficult to add it.

What happens if you don't specify the encoding in the connection string?

IMO, sending an error message without exploding is a feat already. Error reporting is the only place in the library with an error=replace policy: I'm glad it works...

**Comment by lyq1996** [comment]
> I assume that, when that error happens on connection, we don't know the server encoding yet.
> 
> I assume that you assume that psycopg would use the encoding specified in the connection string. It doesn't: that encoding is passed to the server and then psycopg reads it back from the connection, but that doesn't happen when the connection is returned bad. I can see if it is not too difficult to add it.
> 
> What happens if you don't specify the encoding in the connection string?
> 
> IMO, sending an error message without exploding is a feat already. Error reporting is the only place in the library with an error=replace policy: I'm glad it works...

Yes, at least error replace works. I also tried psycopg2, the error message is empty and pgcode, pgerror is None if connection was bad.

If I don't specify the encoding in the connection string, the error message from server still garbled when connection bad, and the error message after the connection is successful also becomes garbled.


I hate GBK :(

**Comment by dvarrazzo** [comment]
> I hate GBK :(

Why is not the database in utf8 encoding?

> If I don't specify the encoding in the connection string, the error message from server still garbled when connection bad, and the error message after the connection is successful also becomes garbled.

That's interesting, it's not what I would have expected.

Would you be able to create a docker image with a server configured to reproduce the issue? That would make simpler to reproduce and test the problem.

**Comment by lyq1996** [comment]
Because the early Win10 uses GBK as the default Chinese encoding, I installed pgsql-server from the official installer without any changes. It uses GBK by default to display error messages (lc_messages archive). If I change lc_messages to utf-8 in postgres.conf, the problem gone..

> Would you be able to create a docker image with a server configured to reproduce the issue? That would make simpler to reproduce and test the problem.

I will build a docker image base on windows server recently.

**Comment by dvarrazzo** [comment]
I was able to reproduce the issue. I should be able to fix the problem. Yes, probably when the error is on connection the server replies in the encoding specified in `lc_messages`. We can check if there is an encoding in the connection string and try to decode the error using it.

**Comment by dvarrazzo** [comment]
Fix will be released in psycopg 3.0.8. and only if the client_encoding is specified in the connection string (otherwise we have no information at all about the encoding of the error messages).

I have debugged the issue on Linux and using the EUC-JP encoding: please test with the maint-3.0 or the master branch to verify that the issue is solved for you too on Windows/BGK. Thank you!

**Comment by lyq1996** [comment]
> Fix will be released in psycopg 3.0.8. and only if the client_encoding is specified in the connection string (otherwise we have no information at all about the encoding of the error messages).
> 
> I have debugged the issue on Linux and using the EUC-JP encoding: please test with the maint-3.0 or the master branch to verify that the issue is solved for you too on Windows/BGK. Thank you!

I tested Windows/GBK with psycopg(3.1.dev0) and it works, thank you!
